src/base/logger.cpp

```diff
@@ -89,9 +91,39 @@ void Logger::addPeer(const QString &ip, const bool blocked, const QString &reaso
     const Log::Peer temp = {m_peerCounter++, QDateTime::currentMSecsSinceEpoch(), ip.toHtmlEscaped(), blocked, reason.toHtmlEscaped()};
     m_peers.push_back(temp);
 
+    locker.unlock();
+
     emit newLogPeer(temp);
 }
 
+const Log::Msg& Logger::getMessage(const int id) const
```

Comment:
Is it safe to return a const reference in this context?
Since both the Logger and the models run on the same thread we don't run the risk of the circular buffer deleting the returned item, correct?

Reply:
>Since both the Logger and the models run on the same thread we don't run the risk of the circular buffer deleting the returned item, correct?

It's incorrect thinking at all. You shouldn't rely on some particular use case(s) when you implement any interface.

Reply:
Then return a temporary instead?
Honestly I can't think of a way around it. I mean: I don't know how to allow external access to the saved items without copying them AND without locking the circular buffer.

Reply:
>I don't know how to allow external access to the saved items without copying them

There is no thread-safe way to do it. So if you want Logger to be thread-safe you can't return references to internal data.
>without locking the circular buffer.

You should lock it every time you read/write it. I.e. you still should lock it here.

Reply:
I know all that.
But since you advised on not actually duplicating the log messages in a vector/list in the Model, what's the alternative here? Let this function return a `Log::Msg` instead?
This function is used by the Model multiple times in row when the Model responds to the various `Qt::roles` via `Model::data()`. Even when you only hover the cursor up and down on the View's items.
Is it a good idea to continually create such short-lived small objects? (I mean the `Log::Msg` returned by the function).

Reply:
>This function is used by the Model multiple times in row when the Model responds to the various `Qt::roles` via `Model::data()`. Even when you only hover the cursor up and down on the View's items.

It isn't the Logger problem. It should have valid interface and implementation according to its own subject domain. It there is a performance problem with LogModel-to-Logger data exchange due to its thread-safity it should be solved in LogModel. Although I remind you once again that you should measure the extent of its impact on performance, because it may be insignificant for GUI. Otherwise the caching data in the LogModel is probably required to solve it.

Reply:
consider the definition of Log::Msg
```
    struct Msg
    {
        int id;
        qint64 timestamp;
        MsgType type;
        QString message;
    };
```
It contains ints with aggregate size not much more than the size of pointer and QString which are internally reference counted, so is this really an issue?

Reply:
I meant by performance implication due to which we're trying to return a reference